### PR TITLE
[WIP] Swap step order

### PR DESCRIPTION
**Pull request by domrigoglioso**
**Patch description**
As in #2042, lr_scheduler.step() was being called before optimizer.step() causing a pytorch warning to appear.

**Testing steps**
Ran `python examples/train_model.py -t convai2 -m transformer/ranker -mf /tmp/tr0 --lr_scheduler invsqrt --warmup_updates 50` and expected to not see any warnings about optimizer.step() after lr_scheduler.step() or other new warnings. 

**Logs**
`python examples/train_model.py -t convai2 -m transformer/ranker -mf /tmp/tr0 --lr_scheduler invsqrt --warmup_updates 50`

Optional arguments output and rest of training omitted
Before changes:
```
[loading fbdialog data:/home/dom/Documents/ParlAI/data/ConvAI2/train_self_original.txt]
[ training... ]
/home/dom/Documents/ParlAI/parlai/core/torch_ranker_agent.py:599: UserWarning: [ Executing train mode with provided inline set of candidates ]
  ''.format(mode)
/home/dom/.local/lib/python3.6/site-packages/torch/optim/lr_scheduler.py:100: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule.See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  "https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate", UserWarning)
/home/dom/Documents/ParlAI/parlai/core/torch_ranker_agent.py:377: UserWarning: Some training metrics are omitted for speed. Set the flag `--train-predict` to calculate train metrics.
  "Some training metrics are omitted for speed. Set the flag "
[ time:2.0s total_exs:11 epochs:0.0 ] {'exs': 11, 'lr': 2.201e-05, 'total_train_updates': 11, 'gnorm': 86.16, 'clip': 1.0, 'examples': 11, 'loss': 40.24, 'mean_loss': 3.658}
[ time:4.0s total_exs:24 epochs:0.0 ] {'exs': 13, 'lr': 4.801e-05, 'total_train_updates': 24, 'gnorm': 73.73, 'clip': 1.0, 'examples': 13, 'loss': 47.29, 'mean_loss': 3.637}
[ time:6.0s total_exs:38 epochs:0.0 ] {'exs': 14, 'lr': 7.6e-05, 'total_train_updates': 38, 'gnorm': 79.09, 'clip': 1.0, 'examples': 14, 'loss': 48.48, 'mean_loss': 3.463}
```

After changes:
```
[loading fbdialog data:/home/dom/Documents/ParlAI/data/ConvAI2/train_self_original.txt]
[ training... ]
/home/dom/Documents/ParlAI/parlai/core/torch_ranker_agent.py:599: UserWarning: [ Executing train mode with provided inline set of candidates ]
  ''.format(mode)
/home/dom/Documents/ParlAI/parlai/core/torch_ranker_agent.py:377: UserWarning: Some training metrics are omitted for speed. Set the flag `--train-predict` to calculate train metrics.
  "Some training metrics are omitted for speed. Set the flag "
[ time:2.0s total_exs:10 epochs:0.0 ] {'exs': 10, 'lr': 2.001e-05, 'total_train_updates': 10, 'gnorm': 106.6, 'clip': 1.0, 'examples': 10, 'loss': 39.58, 'mean_loss': 3.958}
[ time:4.0s total_exs:24 epochs:0.0 ] {'exs': 14, 'lr': 4.801e-05, 'total_train_updates': 24, 'gnorm': 106.0, 'clip': 1.0, 'examples': 14, 'loss': 65.18, 'mean_loss': 4.655}
[ time:6.0s total_exs:38 epochs:0.0 ] {'exs': 14, 'lr': 7.6e-05, 'total_train_updates': 38, 'gnorm': 90.92, 'clip': 1.0, 'examples': 14, 'loss': 53.51, 'mean_loss': 3.822}


```

**Other information**
N/A

**Data tests (if applicable)**
N/A

### CODE REVIEW

**Comment by stephenroller** [comment]
fyi the lines around gradient clipping *must* occur before optimizer.step. Probably you want to move lr_scheduler.step instead.

I think there might be a small issue around the *very* first movement though.

**Comment by stephenroller** [comment]
Tbh we should have a test but I think this is probably okay. I'm just worried about that very first call to `optimizer.step()`.

**Comment by domrigoglioso** [comment]
@stephenroller What would sort of thing would the unit test try to test for?

**Comment by stephenroller** [comment]
That the learning rate is exactly what it should be at steps 0, 1, and 10 or something. Particularly with `--lr-scheduler invsqrt` for example.

**Comment by stephenroller** [comment]
@hadasah when you work on the other LR task, please write some unit tests for the schedulers (isolated) and one integration test.

**Comment by klshuster** [comment]
Just a note that @domrigoglioso spoke with me offline and indicated he will be writing a test for this

**Comment by domrigoglioso** [comment]
I just pushed that unit test, it just checks the invsqrt learning rate is right at step 9 and step 16.

**Comment by klshuster** [comment]
Awesome, looks like this test passes so feel free to merge
